python/ray/tests/test_node_provider_availability_tracker.py

```diff
@@ -0,0 +1,200 @@
+    cur_time = 30
+
+    tracker.update_node_availability("my-node-a", 30, exc)
+
+    # This is after B expires
+    cur_time = 60 * 30 + 11
+    assert len(tracker.summary().node_availabilities) == 2
+
+    summary = tracker.summary()
+
+    assert "my-node-a" in summary.node_availabilities
+    assert "my-node-c" in summary.node_availabilities
+    assert "my-node-b" not in summary.node_availabilities
+
+    assert summary.node_availabilities["my-node-a"].node_type == "my-node-a"
+    assert summary.node_availabilities["my-node-a"].last_checked_timestamp == 30
+
+    assert summary.node_availabilities["my-node-c"].node_type == "my-node-c"
+    assert summary.node_availabilities["my-node-c"].last_checked_timestamp == 20
+
+
+def test_summary(tracker: NodeProviderAvailabilityTracker):
+    exc = NodeLaunchException("DontFeelLikeIt", "This seems like a lot of work.")
+    tracker.update_node_availability("my-node-a", 0, exc)
+
+    tracker.update_node_availability("my-node-b", 1, None)
+
+    summary = tracker.summary()
+
+    expected = {
+        "my-node-a": NodeAvailabilityRecord(
+            node_type="my-node-a",
+            is_available=False,
+            last_checked_timestamp=0,
+            unavailable_node_information=UnavailableNodeInformation(
+                category="DontFeelLikeIt", description="This seems like a lot of work."
+            ),
+        ),
+        "my-node-b": NodeAvailabilityRecord(
+            node_type="my-node-b",
+            is_available=True,
+            last_checked_timestamp=1,
+            unavailable_node_information=None,
+        ),
+    }
+
+    assert summary.node_availabilities == expected
+
+
+def get_timestamp(hour: int, minute: int, second: int, microsecond: int) -> float:
+    # Year, month, day don't just need to be filled out consistently.
+    dt = datetime.datetime(
+        year=2012,
+        month=12,
+        day=21,
+        hour=hour,
+        minute=minute,
+        second=second,
+        microsecond=microsecond,
+    )
+    return dt.timestamp()
+
+
+def test_summary_from_dict():
+    orig = NodeAvailabilitySummary(
+        node_availabilities={
+            "my-node-a": NodeAvailabilityRecord(
+                node_type="my-node-a",
+                is_available=False,
+                last_checked_timestamp=0,
+                unavailable_node_information=UnavailableNodeInformation(
+                    category="DontFeelLikeIt",
+                    description="This seems like a lot of work.",
+                ),
+            ),
+            "my-node-b": NodeAvailabilityRecord(
+                node_type="my-node-b",
+                is_available=True,
+                last_checked_timestamp=1,
+                unavailable_node_information=None,
+            ),
+        }
+    )
+
+    to_dict = dataclasses.asdict(orig)
+
+    remarshalled = NodeAvailabilitySummary.from_fields(**to_dict)
+
+    assert orig == remarshalled
+
+
```

Comment:
Could we add a test that checks in some way that things are unaffected for node providers which do not emit NodeLaunchExceptions?